File: Examples/UsageWithPython/ex05_SampleBuilder/sample_builder.py
#
# extending C++ ISampleBuilder class in python
#
import sys, os
from numpy import linspace, array, zeros, pi, sqrt
import pylab
from matplotlib.colors import LogNorm
import matplotlib.pyplot as plt

sys.path.append(os.path.abspath(
                os.path.join(os.path.split(__file__)[0],
                '..', '..', '..', 'lib')))

from libScattCore import *


#-----------------------------------------------------------
# Cylinders with interference
# Function constructs sample depending on three parameters
#-----------------------------------------------------------
def BuildSample(m_layer_thickness, m_cylinder_height, m_cylinder_radius):
    n_particle = complex(1.0-5e-5, 2e-8)
    p_air_material = MaterialManager.instance().addHomogeneousMaterial("Air", 1.0, 0.0)
    p_layer_material = MaterialManager.instance().addHomogeneousMaterial("Layer", 1.0-2e-6, 2e-8)
    p_substrate_material = MaterialManager.instance().addHomogeneousMaterial("Substrate", 1.0-2e-6, 2e-8)

    air_layer = Layer(p_air_material)
    middle_layer = Layer(p_layer_material, m_layer_thickness)
    substrate_layer = Layer(p_substrate_material)

    interference_funtion = InterferenceFunction1DParaCrystal(20.0*nanometer, 7*nanometer, 1e7*nanometer)
    ff_cylinder = FormFactorCylinder(m_cylinder_height, m_cylinder_radius)
    nano_particle = Particle(n_particle, ff_cylinder )
    particle_collection = ParticleDecoration()
    particle_collection.addParticle(nano_particle)
    particle_collection.addInterferenceFunction(interference_funtion)
    layer_with_particles = LayerDecorator(air_layer, particle_collection)

    p_multi_layer = MultiLayer()
    p_multi_layer.addLayer(layer_with_particles)
    p_multi_layer.addLayer(middle_layer)
    p_multi_layer.addLayer(substrate_layer)
    return p_multi_layer


# ---------------------------------------------------------------
# defining SampleBuilder class which will build samples for us
# 
# ---------------------------------------------------------------
class MySampleBuilder(ISampleBuilder):
    def __init__(self):
        ISampleBuilder.__init__(self)
        print "MySampleBuilder ctor"
        self.layer_thickness = 10.0*nanometer
        self.cylinder_height = 5.0*nanometer
        self.cylinder_radius = 5.0*nanometer
        self.sample = ISample() # this should be the  class variable to increment reference counter and hold object created in buildSample while C++ processes it
        #self.getParameterPool().registerParameter("layer_thickness", self.layer_thickness)

    def buildSample(self):
        self.sample = BuildSample(self.layer_thickness, self.cylinder_height, self.cylinder_radius)
        #sample.print_structure()
        return self.sample


# ---------------------------------------------------------------
# building sample, experiment, run calculations and plot output
# ---------------------------------------------------------------
def main():
  print "hello"
  builder = MySampleBuilder()
  sample = builder.buildSample()

  #sample.print_structure()

  experiment = GISASExperiment()
  experiment.setDetectorParameters(100, 0.0*degree, 2.0*degree,100 , 0.0*degree, 2.0*degree)
  experiment.setBeamParameters(1.0*angstrom, -0.2*degree, 0.0*degree)
  experiment.setBeamIntensity(1e10)
  experiment.setSampleBuilder(builder)
  experiment.runSimulation()

  x = GetOutputDataAxis(experiment, 0)
  y = GetOutputDataAxis(experiment, 1)
  output = GetOutputData(experiment)

  x = x*180./3.1415926
  y = y*180./3.1415926

  fg = pylab.figure()
  X,Y = pylab.meshgrid(x,y)
  pylab.pcolormesh(Y,X,output,  norm=LogNorm(0.1, 1e+04) )
  pylab.ylim([0,y.max()])
  pylab.xlim([0,x.max()])
  pylab.colorbar()
  pylab.show()


#-------------------------------------------------------------
# main()
#-------------------------------------------------------------
if __name__ == '__main__':
  main()
  #pylab.ion()

  #fig1 = pylab.figure()
  #fig1.add_subplot(111).plot([1,2],[3,4])
  ##pylab.draw()

  #fig2 = pylab.figure()
  #fig2.add_subplot(111).plot([5,6],[10,9])
  ##pylab.draw()

  #pylab.show()

File: Tests/FunctionalTests/TestFit/TestFit02/SampleBuilder.cpp
#include "SampleBuilder.h"
#include "MultiLayer.h"
#include "ParticleLayout.h"
#include "InterferenceFunctions.h"
#include "FormFactorCylinder.h"
#include "Units.h"
#include "Materials.h"
#include "MinimizerFactory.h"
#include "FormFactorPrism3.h"
#include "MathFunctions.h"
#include "Utils.h"

using namespace FunctionalTests;

// ----------------------------------------------------------------------------
// Sample builder to build mixture of cylinders and prisms on top of substrate
// 5 parameters
// ----------------------------------------------------------------------------
SampleBuilder::SampleBuilder()
    : m_cylinder_height(5.0*Units::nanometer)
    , m_cylinder_radius(5.0*Units::nanometer)
    , m_prism3_length(10.0*Units::nanometer)
    , m_prism3_height(5.0*Units::nanometer)
    , m_cylinder_ratio(0.2)
{
    init_parameters();
}

// registering parameters in parameter pool
void SampleBuilder::init_parameters()
{
    clearParameterPool();
    registerParameter("cylinder_height", &m_cylinder_height);
    registerParameter("cylinder_radius", &m_cylinder_radius);
    registerParameter("prism3_length", &m_prism3_length);
    registerParameter("prism3_height", &m_prism3_height);
    registerParameter("cylinder_ratio", &m_cylinder_ratio);
}

// build sample from registered parameters
ISample *SampleBuilder::buildSample() const
{
    MultiLayer *multi_layer = new MultiLayer();
    HomogeneousMaterial air_material("Air", 0.0, 0.0);
    HomogeneousMaterial substrate_material("Substrate", 6e-6, 2e-8);
    Layer air_layer(air_material);
    Layer substrate_layer(substrate_material);

    ParticleLayout particle_layout;
    complex_t n_particle(1.0-6e-4, 2e-8);
    HomogeneousMaterial particle_material("Particle", n_particle);

    FormFactorCylinder ff_cylinder(m_cylinder_radius, m_cylinder_height);
    FormFactorPrism3 ff_prism3(m_prism3_length, m_prism3_height);

    particle_layout.addParticle(new Particle(particle_material, ff_cylinder),0.0, m_cylinder_ratio);
    particle_layout.addParticle(new Particle(particle_material, ff_prism3), 0.0, 1.0 - m_cylinder_ratio);
    particle_layout.addInterferenceFunction(new InterferenceFunctionNone());

    air_layer.addLayout(particle_layout);

    multi_layer->addLayer(air_layer);
    multi_layer->addLayer(substrate_layer);
    return multi_layer;
}

